### PR TITLE
Added binary impl of Cursor trait.

### PR DESCRIPTION
**Pull request by zslayton**
*Description of changes:*

Initial binary implementation of the `Cursor` trait.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/binary/cursor.rs [hunk 1]
```diff
@@ -0,0 +1,1010 @@
...
+    // Information about the value on which the cursor is currently sitting
+    value: CursorValue,
+    // All of the values into which the cursor has stepped. Empty at the top level.
+    parents: Vec<CursorValue>,
+}
+
+/// Verifies that the current value is of the expected type and that the bytes representing that
+/// value have not yet been consumed from the data source. This macro is called by the
+/// BinaryCursor#read_{typename} methods.
+macro_rules! read_safety_checks {
```
Comment (zslayton) [inline src/binary/cursor.rs:99]: Making this a macro allowed me to:
* Repeat the same `if` tests much more concisely throughout the code.
* Perform early returns from the call sites without requiring `if ... {return}` scaffolding.

Comment (therapon) [reply src/binary/cursor.rs:99]: minor: naming of the macro seems off to me, perhaps, `read_type` or `read_val_of_type`

File: src/binary/cursor.rs [hunk 2]
```diff
@@ -0,0 +1,1010 @@
...
+        IonTypeCode,
+    },
+    data_source::IonDataSource,
+    result::{decoding_error, illegal_operation, illegal_operation_raw, IonResult},
+    types::{IonType, SymbolId},
+};
+
+/// Information about the value over which the Cursor is currently positioned.
+#[derive(Clone, Debug)]
+struct CursorValue {
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/binary/cursor.rs:26]: Per our offline conversation, this type is also used to represent parent values stored in the stack that's managed by calls to `step_in` and `step_out`. I've renamed it to `EncodedValue`.

File: src/binary/cursor.rs [hunk 3]
```diff
@@ -0,0 +1,1010 @@
...
+    }
+}
+
+// A low-level reader that offers no validation or symbol management.
+// It can only move and return the current value.
+pub struct BinaryIonCursor<R>
+where
+    R: IonDataSource,
+{
+    // The file, socket, array, or other BufRead implementor containing binary Ion bytes
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/binary/cursor.rs:64]: Sorry, this was muddy phrasing on my part. The `IonDataSource` trait is an extension trait, implemented for all values who implement `BufRead`. I've removed the use of the trait name to make it less confusing.

File: src/binary/cursor.rs [hunk 4]
```diff
@@ -0,0 +1,1010 @@
...
+    };
+}
+
+impl<R: IonDataSource> Cursor<R> for BinaryIonCursor<R> {
+    fn ion_version(&self) -> (u8, u8) {
+        self.cursor.ion_version
+    }
+
+    fn next(&mut self) -> IonResult<Option<StreamItem>> {
+        // If the cursor is partway through reading a value, skip the rest of that value.
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/binary/cursor.rs:124]: Old comment, good catch. It skips the rest of the value no matter what.

File: src/binary/cursor.rs [hunk 5]
```diff
@@ -0,0 +1,1010 @@
...
+    fn read_i64(&mut self) -> IonResult<Option<i64>> {
+        read_safety_checks!(self, IonType::Integer);
+
+        let magnitude = self.read_value_as_uint()?.value();
+
+        use self::IonTypeCode::*;
+        let value = match self.cursor.value.header.ion_type_code {
+            PositiveInteger => magnitude as i64,
+            NegativeInteger => -(magnitude as i64),
+            _ => unreachable!("The Ion Type Code must be one of the above to reach this point."),
```
[the remark this answers is not in the document]
Comment (zslayton) [reply src/binary/cursor.rs:228]: You'll never see the error message. It's `unreachable!` 😛 Fixed, thanks.

**Pull request merged**